File: src/hackerrank/java/ChainingLetters.java
package hackerrank.java;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class ChainingLetters {

    public static void main(String[] args) {
        String[] arr = {"A>I", "S>P", "P>A", "I>N"};
        System.out.println(chainingLetters(arr));
    }

    private static String chainingLetters(String[] array) {

        Map<String, String> map = new HashMap<>();

        Set<String> set = new HashSet<>();

        for (int i = 0; i <= array.length; i++) {
            String[] pair = array[i].split(">");
            String first = pair[0];
            String second = pair[1];

            if (set.contains(first)) {
                set.remove(first);

            } else {
                set.add(first);

            }
            if (set.contains(second)) {
                set.remove(second);

            } else {

                set.add(second);
            }
            map.put(first, second);
        }
        String firstPosition = "";
        for (int i = 0; i < array.length; i++) {
            String[] pair = array[i].split(">");
            String first = pair[0];

            if (set.contains(first)) {
                firstPosition = first;
                break;
            }
        }

        String result = firstPosition;

        for (int i = 0; i < array.length; i++) {
            String novi = map.get(firstPosition);
            result += novi;
            firstPosition = novi;
        }

        return result;
    }
}

File: src/hackerrank/kotlin/ChainingLetters.kt
package hackerrank.kotlin

fun main(args: Array<String>) {
    chainingLetters(arrayOf("A>I", "S>P", "P>A", "I>N"))
}

fun chainingLetters(arr: Array<String>) {

    val map = mutableMapOf<String, String>()
    arr.forEach {
        val group = it.split(">")
        val first = group[0]
        val second = group[1]
        map[first] = second
    }

    var firstLetter = map.keys.first { !map.entries.groupBy { it.value }.containsKey(it) }

    var result = firstLetter

    arr.forEach {
        val next = map[firstLetter]
        result+= next
        firstLetter = next.toString()
    }
    println(result)
}